src/Workspaces/Core/Portable/CodeFixesAndRefactorings/DocumentBasedFixAllProviderHelpers.cs

```diff
@@ -30,124 +28,113 @@ internal static class DocumentBasedFixAllProviderHelpers
         ImmutableArray<TFixAllContext> fixAllContexts,
         IProgress<CodeAnalysisProgress> progressTracker,
         string progressTrackerDescription,
-        Func<TFixAllContext, IProgress<CodeAnalysisProgress>, Task<Dictionary<DocumentId, (SyntaxNode? node, SourceText? text)>>> getFixedDocumentsAsync)
+        Func<TFixAllContext, Action<(DocumentId documentId, (SyntaxNode? node, SourceText? text))>, Task> getFixedDocumentsAsync)
         where TFixAllContext : IFixAllContext
     {
+        var cancellationToken = originalFixAllContext.CancellationToken;
+
         progressTracker.Report(CodeAnalysisProgress.Description(progressTrackerDescription));
 
         var solution = originalFixAllContext.Solution;
 
-        // For code fixes, we have 3 pieces of work per project.  Computing diagnostics, computing fixes, and applying fixes.
-        // For refactorings, we have 2 pieces of work per project.  Computing refactorings, and applying refactorings.
-        var fixAllKind = originalFixAllContext.State.FixAllKind;
-        var workItemCount = fixAllKind == FixAllKind.CodeFix ? 3 : 2;
-        progressTracker.AddItems(fixAllContexts.Length * workItemCount);
+        // One work item for each context.
+        progressTracker.AddItems(fixAllContexts.Length);
```

Comment:
the math computation was originally wrong. it was not hte case that we did 3 items for fixes per context, and 2 per refactoring per context. Because this was just broken. i simplified this. the computation of progress length is done here, and the updating happens here as well. this is my preference for progress as having length computation and item-finish calls separated means it's too easy to get wrong.